### PR TITLE
[ARCTIC-260][Core]Support loading Iceberg catalog

### PR DESCRIPTION
**Pull request by zhoujinsong**
## Why are the changes needed?

Subtask of #260.
Support loading iceberg catalog.

## Brief change log

  - *Add BaseIcebergCatalog implementation of ArcticCatalog for native iceberg usage*
  - *CatalogLoader support loading BaseIcebergCatalog*
  - *Add authrozation for BaseIcebergCatalog*

## How was this patch tested?
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

## Documentation

  - Does this pull request introduces a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

### CODE REVIEW

File: core/src/main/java/com/netease/arctic/catalog/CatalogLoader.java [hunk 1]
```diff
@@ -93,18 +101,43 @@ public static ArcticCatalog load(AmsClient client, String catalogName, Map<Strin
...
+      CatalogUtil.mergeCatalogProperties(catalogMeta, props);
       String catalogImpl;
+      Set<TableFormat> tableFormats = CatalogUtil.tableFormats(catalogMeta);
+      Preconditions.checkArgument(tableFormats.size() == 1, "Catalog support only one table format now.");
+      TableFormat tableFormat = tableFormats.iterator().next();
       switch (type) {
         case CATALOG_TYPE_HADOOP:
-          catalogImpl = HADOOP_CATALOG_IMPL;
+          Preconditions.checkArgument(tableFormat.equals(TableFormat.ICEBERG),
+              "Hadoop catalog support hive table only.");
```
Comment (hzluting) [inline core/src/main/java/com/netease/arctic/catalog/CatalogLoader.java:112]: The check condition and the error message don't match

**Review by hzluting — APPROVED** [review]

File: core/src/main/java/com/netease/arctic/catalog/CatalogLoader.java [hunk 2]
```diff
@@ -93,18 +101,43 @@ public static ArcticCatalog load(AmsClient client, String catalogName, Map<Strin
     try {
       CatalogMeta catalogMeta = client.getCatalog(catalogName);
       String type = catalogMeta.getCatalogType();
+      CatalogUtil.mergeCatalogProperties(catalogMeta, props);
       String catalogImpl;
+      Set<TableFormat> tableFormats = CatalogUtil.tableFormats(catalogMeta);
+      Preconditions.checkArgument(tableFormats.size() == 1, "Catalog support only one table format now.");
+      TableFormat tableFormat = tableFormats.iterator().next();
       switch (type) {
```
Comment (baiyangtx) [inline core/src/main/java/com/netease/arctic/catalog/CatalogLoader.java:109]: 这边在实现上，是不是不要在代码层面复用 type 字段了。 v0.4 就直接引入  metastore,  table-format 两个新字段。 v0.3 的还走 type 逻辑。 这样代码上更清晰一些。
在兼容性上，保证 v0.4 接口还有 type 字段，v0.5 开始就不返回 type 字段了？

File: core/src/main/java/com/netease/arctic/catalog/BaseIcebergCatalog.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+        meta.getCatalogProperties(), tableMetaStore.getConfiguration());
+  }
+
+  @Override
+  public List<String> listDatabases() {
+    if (icebergCatalog instanceof SupportsNamespaces) {
+      return tableMetaStore.doAs(() -> ((SupportsNamespaces)icebergCatalog).listNamespaces(Namespace.empty()).stream()
+          .map(namespace -> namespace.level(0)).distinct().collect(Collectors.toList()));
+    } else {
+      throw new UnsupportedOperationException(String.format("Iceberg catalog: %s do now implement SupportsNamespaces",
```
Comment (Aireed) [inline core/src/main/java/com/netease/arctic/catalog/BaseIcebergCatalog.java:72]: `throw new UnsupportedOperationException(String.format("Iceberg catalog: %s do now implement SupportsNamespaces",`

'do now' should be does not ?

File: core/src/main/java/com/netease/arctic/catalog/BaseIcebergCatalog.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+  @Override
+  public void createDatabase(String databaseName) {
+    if (icebergCatalog instanceof SupportsNamespaces) {
+      tableMetaStore.doAs(() -> {
+        ((SupportsNamespaces)icebergCatalog).createNamespace(Namespace.of(databaseName));
+        return null;
+      });
+    } else {
+      throw new UnsupportedOperationException(String.format("Iceberg catalog: %s do now implement SupportsNamespaces",
```
Comment (Aireed) [inline core/src/main/java/com/netease/arctic/catalog/BaseIcebergCatalog.java:85]: do now --> does not

File: core/src/main/java/com/netease/arctic/catalog/CatalogLoader.java (hunk 2 above)
Comment (zhoujinsong) [reply core/src/main/java/com/netease/arctic/catalog/CatalogLoader.java:109]: catalog `type` is the same with `metastore` in AMS Dashboard. 
I prefer do not add a `metastore` property in database and thrift argument.

**Review by baiyangtx — APPROVED** [review]

File: core/src/main/java/com/netease/arctic/catalog/BaseIcebergCatalog.java [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+  @Override
+  public void dropDatabase(String databaseName) {
+    if (icebergCatalog instanceof SupportsNamespaces) {
+      tableMetaStore.doAs(() -> {
+        ((SupportsNamespaces)icebergCatalog).dropNamespace(Namespace.of(databaseName));
+        return null;
+      });
+    } else {
+      throw new UnsupportedOperationException(String.format("Iceberg catalog: %s do now implement SupportsNamespaces",
```
Comment (Aireed) [inline core/src/main/java/com/netease/arctic/catalog/BaseIcebergCatalog.java:98]: throw new UnsupportedOperationException(String.format("Iceberg catalog: %s do now implement SupportsNamespaces",

'do now' should be does not ?

**Pull request merged**